### PR TITLE
fix: support keyboard navigation (arrow key up/down) in grid mode

### PR DESCRIPTION
**Pull request by quanho**
## Motivation

To fix https://github.com/qlik-oss/sn-list-objects/issues/277.

Screenshot 2023-04-21 at 19 40 11

## Requirements checklist

- [ ] Api specification
  - [ ] Ran `yarn spec`
    - [ ] No changes **_OR_** API changes has been formally approved
- [ ] Unit/Component test coverage
- [ ] Correct PR title for the changes (fix, chore, feat)

When build and tests have passed:

- [ ] Add code reviewers, for example @qlik-oss/nebula-core

### CODE REVIEW

File: apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js [hunk 1]
```diff
@@ -11,10 +11,69 @@ export default function getRowsKeyboardNavigation({
...
+      keyCode === KEYS.ARROW_RIGHT ||
+      !(typeof rowIndex === 'number' && typeof columnIndex === 'number')
+    ) {
+      const parentElm = elm && elm.parentElement[next ? 'nextElementSibling' : 'previousElementSibling'];
+      return parentElm && parentElm.querySelector('[role]');
+    }
+    const gridElm = elm && elm.parentElement.parentElement;
+    if (gridElm && gridElm.childElementCount) {
+      const nextIndex = findNextIndex(keyCode, gridElm.childElementCount);
+      const nextElm = elm && elm.parentElement.parentElement.children[nextIndex];
```
Comment (johanlahti) [inline apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js:73]:
```suggestion
      const nextElm = elm?.parentElement.parentElement.children[nextIndex];
```

File: apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js [hunk 2]
```diff
@@ -11,10 +11,69 @@ export default function getRowsKeyboardNavigation({
...
+  };
+
+  const getElement = (keyCode, elm, next = false) => {
+    if (
+      keyCode === KEYS.ARROW_LEFT ||
+      keyCode === KEYS.ARROW_RIGHT ||
+      !(typeof rowIndex === 'number' && typeof columnIndex === 'number')
+    ) {
+      const parentElm = elm && elm.parentElement[next ? 'nextElementSibling' : 'previousElementSibling'];
+      return parentElm && parentElm.querySelector('[role]');
```
Comment (johanlahti) [inline apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js:68]:
```suggestion
      const parentElm = elm?.parentElement[next ? 'nextElementSibling' : 'previousElementSibling'];
      return parentElm?.querySelector('[role]');
```

File: apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js [hunk 3]
```diff
@@ -11,10 +11,69 @@ export default function getRowsKeyboardNavigation({
...
+    }
+    const gridElm = elm && elm.parentElement.parentElement;
+    if (gridElm && gridElm.childElementCount) {
+      const nextIndex = findNextIndex(keyCode, gridElm.childElementCount);
+      const nextElm = elm && elm.parentElement.parentElement.children[nextIndex];
+      return nextElm && nextElm.querySelector('[role]');
```
Comment (johanlahti) [inline apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js:74]:
```suggestion
      return nextElm?.querySelector('[role]');
```

File: apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js [hunk 4]
```diff
@@ -11,10 +11,69 @@ export default function getRowsKeyboardNavigation({
...
+    }
+    return undefined;
```
Comment (johanlahti) [inline apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js:76]: What happens when we are _not_ in grid mode (i.e. singleColumn mode)? I would expect these to have same effect:
- up, left -> go up
- down, right -> go down

Or?

File: apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js [hunk 5]
```diff
@@ -11,10 +11,69 @@ export default function getRowsKeyboardNavigation({
...
+    if (
+      keyCode === KEYS.ARROW_LEFT ||
+      keyCode === KEYS.ARROW_RIGHT ||
+      !(typeof rowIndex === 'number' && typeof columnIndex === 'number')
+    ) {
+      const parentElm = elm && elm.parentElement[next ? 'nextElementSibling' : 'previousElementSibling'];
+      return parentElm && parentElm.querySelector('[role]');
+    }
+    const gridElm = elm && elm.parentElement.parentElement;
+    if (gridElm && gridElm.childElementCount) {
```
Comment (johanlahti) [inline apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js:71]:
```suggestion
    if (gridElm?.childElementCount) {
```

File: apis/nucleus/src/components/listbox/components/ListBoxRowColumn/ListBoxRowColumn.jsx [hunk 6]
```diff
@@ -102,7 +102,17 @@ function RowColumn({ index, rowIndex, columnIndex, style, data }) {
...
+      focusListItems,
+      keyboard,
+      isModal,
+      rowCount,
+      columnCount,
+      rowIndex,
+      columnIndex,
+      layoutOrder,
+    }),
     [actions, keyboard?.innerTabStops]
```
Comment (T-Wizard) [inline apis/nucleus/src/components/listbox/components/ListBoxRowColumn/ListBoxRowColumn.jsx:116]: you probably should add the new parameters as dependencies here
(to get an updated callback if they change)

File: apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js
Comment (T-Wizard) [inline apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js:1]: try add unit test for the new logic in this file

Comment (quanho) [reply apis/nucleus/src/components/listbox/interactions/keyboard-navigation/keyboard-nav-rows.js:1]: Added!

File: apis/nucleus/src/components/listbox/components/ListBoxRowColumn/ListBoxRowColumn.jsx [hunk 7]
```diff
@@ -102,8 +102,18 @@ function RowColumn({ index, rowIndex, columnIndex, style, data }) {
...
+      focusListItems,
+      keyboard,
+      isModal,
+      rowCount,
+      columnCount,
+      rowIndex,
+      columnIndex,
+      layoutOrder,
+    }),
+    [actions, keyboard?.innerTabStops, rowCount, columnCount, rowIndex, columnIndex, layoutOrder]
```
Comment (johanlahti) [inline apis/nucleus/src/components/listbox/components/ListBoxRowColumn/ListBoxRowColumn.jsx:116]: Don't know if needed… but might have to add `focusListItems` too:
```suggestion
    [actions, focusListItems, keyboard?.innerTabStops, rowCount, columnCount, rowIndex, columnIndex, layoutOrder]
```

**Review by johanlahti — APPROVED** [review]

**Pull request merged**